File: ruby/lib/svix/models/poller_v2_message_out.rb
# frozen_string_literal: true
# This file is @generated
require "json"

module Svix
  # The MessageOut equivalent of polling endpoint
  class PollerV2MessageOut
    attr_accessor :offset
    attr_accessor :headers
    # Optional unique identifier for the message
    attr_accessor :event_id
    # The event type's name
    attr_accessor :event_type
    attr_accessor :payload
    # List of free-form identifiers that endpoints can filter by
    attr_accessor :channels
    # The Message's ID.
    attr_accessor :id
    attr_accessor :timestamp
    attr_accessor :tags
    attr_accessor :deliver_at

    ALL_FIELD ||= [
      "offset",
      "headers",
      "event_id",
      "event_type",
      "payload",
      "channels",
      "id",
      "timestamp",
      "tags",
      "deliver_at"
    ].freeze
    private_constant :ALL_FIELD

    def initialize(attributes = {})
      unless attributes.is_a?(Hash)
        fail(ArgumentError, "The input argument (attributes) must be a hash in `Svix::PollerV2MessageOut` new method")
      end

      attributes.each do |k, v|
        unless ALL_FIELD.include?(k.to_s)
          fail(ArgumentError, "The field #{k} is not part of Svix::PollerV2MessageOut")
        end

        instance_variable_set("@#{k}", v)
        instance_variable_set("@__#{k}_is_defined", true)
      end
    end

    def self.deserialize(attributes = {})
      attributes = attributes.transform_keys(&:to_s)
      attrs = Hash.new
      attrs["offset"] = attributes["offset"]
      attrs["headers"] = attributes["headers"]
      attrs["event_id"] = attributes["eventId"]
      attrs["event_type"] = attributes["eventType"]
      attrs["payload"] = attributes["payload"]
      attrs["channels"] = attributes["channels"]
      attrs["id"] = attributes["id"]
      attrs["timestamp"] = DateTime.rfc3339(attributes["timestamp"]).to_time
      attrs["tags"] = attributes["tags"]
      attrs["deliver_at"] = DateTime.rfc3339(attributes["deliverAt"]).to_time if attributes["deliverAt"]
      new(attrs)
    end

    def serialize
      out = Hash.new
      out["offset"] = Svix::serialize_primitive(@offset) if @offset
      out["headers"] = Svix::serialize_primitive(@headers) if @headers
      out["eventId"] = Svix::serialize_primitive(@event_id) if @event_id
      out["eventType"] = Svix::serialize_primitive(@event_type) if @event_type
      out["payload"] = Svix::serialize_primitive(@payload) if @payload
      out["channels"] = Svix::serialize_primitive(@channels) if @channels
      out["id"] = Svix::serialize_primitive(@id) if @id
      out["timestamp"] = Svix::serialize_primitive(@timestamp) if @timestamp
      out["tags"] = Svix::serialize_primitive(@tags) if @tags
      out["deliverAt"] = Svix::serialize_primitive(@deliver_at) if @deliver_at
      out
    end

    # Serializes the object to a json string
    # @return String
    def to_json
      JSON.dump(serialize)
    end
  end
end

File: php/src/Models/PollerV2MessageOut.php
<?php

// this file is @generated
declare(strict_types=1);

namespace Svix\Models;

/** The MessageOut equivalent of polling endpoint */
class PollerV2MessageOut implements \JsonSerializable
{
    private array $setFields = [];

    /**
     * @param array<string, string>|null $headers
     * @param string|null                $eventId   Optional unique identifier for the message
     * @param string                     $eventType The event type's name
     * @param list<string>|null          $channels  List of free-form identifiers that endpoints can filter by
     * @param string                     $id        the Message's ID
     * @param list<string>|null          $tags
     */
    private function __construct(
        public readonly int $offset,
        public readonly string $eventType,
        public readonly array $payload,
        public readonly string $id,
        public readonly \DateTimeImmutable $timestamp,
        public readonly ?array $headers = null,
        public readonly ?string $eventId = null,
        public readonly ?array $channels = null,
        public readonly ?array $tags = null,
        public readonly ?\DateTimeImmutable $deliverAt = null,
        array $setFields = [],
    ) {
        $this->setFields = $setFields;
    }

    /**
     * Create an instance of PollerV2MessageOut with required fields.
     */
    public static function create(
        int $offset,
        string $eventType,
        array $payload,
        string $id,
        \DateTimeImmutable $timestamp,
    ): self {
        return new self(
            offset: $offset,
            headers: null,
            eventId: null,
            eventType: $eventType,
            payload: $payload,
            channels: null,
            id: $id,
            timestamp: $timestamp,
            tags: null,
            deliverAt: null,
            setFields: ['offset' => true, 'eventType' => true, 'payload' => true, 'id' => true, 'timestamp' => true]
        );
    }

    public function withHeaders(?array $headers): self
    {
        $setFields = $this->setFields;
        $setFields['headers'] = true;

        return new self(
            offset: $this->offset,
            headers: $headers,
            eventId: $this->eventId,
            eventType: $this->eventType,
            payload: $this->payload,
            channels: $this->channels,
            id: $this->id,
            timestamp: $this->timestamp,
            tags: $this->tags,
            deliverAt: $this->deliverAt,
            setFields: $setFields
        );
    }

    public function withEventId(?string $eventId): self
    {
        $setFields = $this->setFields;
        $setFields['eventId'] = true;

        return new self(
            offset: $this->offset,
            headers: $this->headers,
            eventId: $eventId,
            eventType: $this->eventType,
            payload: $this->payload,
            channels: $this->channels,
            id: $this->id,
            timestamp: $this->timestamp,
            tags: $this->tags,
            deliverAt: $this->deliverAt,
            setFields: $setFields
        );
    }

    public function withChannels(?array $channels): self
    {
        $setFields = $this->setFields;
        $setFields['channels'] = true;

        return new self(
            offset: $this->offset,
            headers: $this->headers,
            eventId: $this->eventId,
            eventType: $this->eventType,
            payload: $this->payload,
            channels: $channels,
            id: $this->id,
            timestamp: $this->timestamp,
            tags: $this->tags,
            deliverAt: $this->deliverAt,
            setFields: $setFields
        );
    }

    public function withTags(?array $tags): self
    {
        $setFields = $this->setFields;
        $setFields['tags'] = true;

        return new self(
            offset: $this->offset,
            headers: $this->headers,
            eventId: $this->eventId,
            eventType: $this->eventType,
            payload: $this->payload,
            channels: $this->channels,
            id: $this->id,
            timestamp: $this->timestamp,
            tags: $tags,
            deliverAt: $this->deliverAt,
            setFields: $setFields
        );
    }

    public function withDeliverAt(?\DateTimeImmutable $deliverAt): self
    {
        $setFields = $this->setFields;
        $setFields['deliverAt'] = true;

        return new self(
            offset: $this->offset,
            headers: $this->headers,
            eventId: $this->eventId,
            eventType: $this->eventType,
            payload: $this->payload,
            channels: $this->channels,
            id: $this->id,
            timestamp: $this->timestamp,
            tags: $this->tags,
            deliverAt: $deliverAt,
            setFields: $setFields
        );
    }

    public function jsonSerialize(): mixed
    {
        $data = [
            'offset' => $this->offset,
            'eventType' => $this->eventType,
            'payload' => $this->payload,
            'id' => $this->id,
            'timestamp' => $this->timestamp->format('c')];

        if (isset($this->setFields['headers'])) {
            $data['headers'] = $this->headers;
        }
        if (isset($this->setFields['eventId'])) {
            $data['eventId'] = $this->eventId;
        }
        if (isset($this->setFields['channels'])) {
            $data['channels'] = $this->channels;
        }
        if (isset($this->setFields['tags'])) {
            $data['tags'] = $this->tags;
        }
        if (isset($this->setFields['deliverAt'])) {
            $data['deliverAt'] = $this->deliverAt->format('c');
        }

        return \Svix\Utils::newStdClassIfArrayIsEmpty($data);
    }

    /**
     * Create an instance from a mixed obj.
     */
    public static function fromMixed(mixed $data): self
    {
        return new self(
            offset: \Svix\Utils::deserializeInt($data, 'offset', true, 'PollerV2MessageOut'),
            headers: \Svix\Utils::getValFromJson($data, 'headers', false, 'PollerV2MessageOut'),
            eventId: \Svix\Utils::deserializeString($data, 'eventId', false, 'PollerV2MessageOut'),
            eventType: \Svix\Utils::deserializeString($data, 'eventType', true, 'PollerV2MessageOut'),
            payload: \Svix\Utils::getValFromJson($data, 'payload', true, 'PollerV2MessageOut'),
            channels: \Svix\Utils::getValFromJson($data, 'channels', false, 'PollerV2MessageOut'),
            id: \Svix\Utils::deserializeString($data, 'id', true, 'PollerV2MessageOut'),
            timestamp: \Svix\Utils::deserializeDt($data, 'timestamp', true, 'PollerV2MessageOut'),
            tags: \Svix\Utils::getValFromJson($data, 'tags', false, 'PollerV2MessageOut'),
            deliverAt: \Svix\Utils::deserializeDt($data, 'deliverAt', false, 'PollerV2MessageOut')
        );
    }

    /**
     * Create an instance from a json string.
     */
    public static function fromJson(string $json): self
    {
        $data = json_decode(json: $json, associative: true, depth: 512, flags: JSON_THROW_ON_ERROR);

        return self::fromMixed($data);
    }
}
